src/Compilers/CSharp/Test/Emit/Emit/EntryPointTests.cs

```diff
@@ -107,6 +107,153 @@ public class D
                 Diagnostic(ErrorCode.ERR_MultipleEntryPoints, "Main"));
         }
 
+        [Fact]
+        public void WRN_SyncAndAsyncEntryPoints_CSharp71()
+        {
+            string source = @"
+using System.Threading.Tasks;
+
+public class C
+{
+  public static async Task Main() { await Task.Delay(1); }
+  public static void Main(string[] a) { System.Console.WriteLine(2); }
+}
+";
+            var compilation = CompileConsoleApp(source, parseOptions: TestOptions.Regular7_1);
+
+            compilation.VerifyDiagnostics(
+                // (6,28): warning CS8892: Method 'C.Main()' will not be used as an entry point because a synchronous entry point 'C.Main(string[])' was found.
+                Diagnostic(ErrorCode.WRN_SyncAndAsyncEntryPoints, "Main").WithArguments("C.Main()", "C.Main(string[])").WithLocation(6, 28));
+        }
+
+        [Fact]
+        public void SyncAndAsyncEntryPointsBeforeAsyncMainFeature_NoDiagnostics()
+        {
+            string source = @"
+using System.Threading.Tasks;
+
+public class C
+{
+  public static async Task Main() { await Task.Delay(1); }
+  public static void Main(string[] a) { System.Console.WriteLine(2); }
+}
+";
+            var compilation = CompileConsoleApp(source, parseOptions: TestOptions.Regular7);
+
+            compilation.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void WRN_SyncAndAsyncEntryPointsCSharpLatest_SyncAndAsync()
+        {
+            string source = @"
+using System.Threading.Tasks;
+
+public class C
+{
+  public static async Task Main() { await Task.Delay(1); }
+  public static void Main(string[] a) { System.Console.WriteLine(2); }
+}
+";
+            var compilation = CompileConsoleApp(source);
+
+            compilation.VerifyDiagnostics(
+                // (6,28): warning CS8892: Method 'C.Main()' will not be used as an entry point because a synchronous entry point 'C.Main(string[])' was found.
+                Diagnostic(ErrorCode.WRN_SyncAndAsyncEntryPoints, "Main").WithArguments("C.Main()", "C.Main(string[])").WithLocation(6, 28));
+        }
+
+        [Fact]
+        public void ERR_MultipleEntryPointsCSharpLatest_TwoSyncAndOneAsync()
+        {
+            string source = @"
+using System.Threading.Tasks;
+
+public class C
+{
+  public static async Task Main() { await Task.Delay(1); }
+  public static void Main(string[] a) { System.Console.WriteLine(2); }
+}
+
+public class D
+{
+  public static void Main() { System.Console.WriteLine(3); }
+}
+";
+            var compilation = CompileConsoleApp(source);
```

Comment:
Consider testing when `mainTypeName: C` is specified and when `mainTypeName: D` is specified. I wonder if the expected result of 1 warning diagnostic occurs when we specify C and 0 diagnostics when we specify D (I could have misinterpreted things, but that is my read of what the flag should do..)